### PR TITLE
[FLINK-25415] implement retrial on connections to Cassandra container

### PR DESCRIPTION
**Pull request by echauchot**
R: @tillrohrmann 
CC: @zentol

### CODE REVIEW

**Comment by echauchot** [comment]
@flinkbot run azure

**Comment by MartijnVisser** [comment]
@flinkbot run azure

**Comment by MartijnVisser** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 1]
```diff
@@ -92,6 +98,12 @@
     @ClassRule
     public static final CassandraContainer CASSANDRA_CONTAINER = createCassandraContainer();
 
+    private static final int MAX_CONNECTION_RETRY = 3;
+    private static final int CONNECTION_RETRY_DELAY = 500;
```
Comment (tillrohrmann) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java:102]: nit ```suggestion
    private static final long CONNECTION_RETRY_DELAY = 500L;
```

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 2]
```diff
@@ -83,7 +88,8 @@
 
 /** IT cases for all cassandra sinks. */
 @SuppressWarnings("serial")
-@Ignore(value = "Flaky test")
+// NoHostAvailableException is raised by Cassandra client under load while connecting to the cluster
+@RetryOnException(times = 3, exception = NoHostAvailableException.class)
```
Comment (tillrohrmann) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java:92]: Why is this needed if we have the retry logic in the setup method? The JIRA reads as if the `NoHostAvailableException` only occurs when the client connects to Cassandra. Can this also happen after the `session` is established?

**Review by tillrohrmann — COMMENTED** [review]
Thanks for creating this PR @echauchot. The changes look good to me. I had a question concerning the `@RetryOnException` annotation.

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 1 above)
Comment (echauchot) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java:102]: sure +1

**Comment by echauchot** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 2 above)
Comment (echauchot) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java:92]: It surprised me also but it is needed because `CassandraCommitter#createResource()` calls `cluster.connect();` and most of `CassandraConnectorITCase` tests rely on `CassandraCommitter`

**Comment by echauchot** [comment]
@tillrohrmann thanks for the prompt review ! I addressed the nit comment and squashed and answered your other comment. PTAL

**Comment by echauchot** [comment]
azure failure was due to a task manager failing in out of memory during an unaligned checkpoint test. It is unrelated to the current PR but I saw that happening several times. Maybe we could raise the memory size of task managers in the CI. WDYT ?

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 2 above)
Comment (tillrohrmann) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java:92]: Thanks for the clarification @echauchot.

**Review by tillrohrmann — APPROVED** [review]
Thanks for the clarification and update of the PR @echauchot. LGTM. Merging this PR once AZP gives green light.

**Comment by echauchot** [comment]
> Thanks for creating this PR @echauchot. The changes look good to me. I had a question concerning the `@RetryOnException` annotation.

I tried to be as quick as possible considering that the test was breaking the CI build more often and that it was disabled

**Comment by echauchot** [comment]
@tillrohrmann CI is green, can we merge the PR ? As usual we will monitor for flakiness for a week or two and see if the PR solves the problem.

**Comment by echauchot** [comment]
@tillrohrmann thanks for merging ! And thanks for backporting to 1.13 and 1.14 !
